### PR TITLE
DRY up BaseSelect2Handler

### PR DESCRIPTION
**Pull request by orangejenny**
Tested locally that both the select2 handler in accounting.software_plan_version_handler.js (on the plan edit page) and the other usages of select2_handler.js (domain sms rates, global sms rates) all appear to still work fine.

`w=1`

@dannyroberts / @nickpell

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 18 errors:

* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 70 - Missing trailing comma. (comma-dangle)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 76 - Missing trailing comma. (comma-dangle)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 107 - Missing trailing comma. (comma-dangle)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 108 - Missing trailing comma. (comma-dangle)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 111 - Missing semicolon. (semi)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 121 - Expected '===' and instead saw '=='. (eqeqeq)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 124 - Expected '===' and instead saw '=='. (eqeqeq)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 151 - 'e' is defined but never used (no-unused-vars)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 156 - Missing semicolon. (semi)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 218 - Missing semicolon. (semi)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 225 - 'BaseSelect2Handler' is not defined. (no-undef)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 236 - 'term' is defined but never used (no-unused-vars)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 238 - Missing trailing comma. (comma-dangle)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 239 - Missing semicolon. (semi)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 249 - Missing semicolon. (semi)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 270 - 'event' is defined but never used (no-unused-vars)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 271 - Expected '===' and instead saw '=='. (eqeqeq)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 278 - 'BaseSelect2Handler' is not defined. (no-undef)

**Comment by orangejenny** [comment]
I should wrap select2_handler.js in an `hqDefine` and then import it in `accounting.software_plan_version_handler.js`. Will do that.

**Comment by dimagimon** [comment]
There are 29 errors:

* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 70 - Missing trailing comma. (comma-dangle)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 76 - Missing trailing comma. (comma-dangle)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 107 - Missing trailing comma. (comma-dangle)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 108 - Missing trailing comma. (comma-dangle)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 111 - Missing semicolon. (semi)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 121 - Expected '===' and instead saw '=='. (eqeqeq)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 124 - Expected '===' and instead saw '=='. (eqeqeq)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 151 - 'e' is defined but never used (no-unused-vars)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 156 - Missing semicolon. (semi)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 218 - Missing semicolon. (semi)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 237 - 'term' is defined but never used (no-unused-vars)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 239 - Missing trailing comma. (comma-dangle)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 240 - Missing semicolon. (semi)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 250 - Missing semicolon. (semi)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 271 - 'event' is defined but never used (no-unused-vars)
* corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js, line 272 - Expected '===' and instead saw '=='. (eqeqeq)
* corehq/apps/hqwebapp/static/hqwebapp/js/select2_handler.js, line 29 - 'selectedData' is defined but never used (no-unused-vars)
* corehq/apps/hqwebapp/static/hqwebapp/js/select2_handler.js, line 42 - 'element' is defined but never used (no-unused-vars)
* corehq/apps/hqwebapp/static/hqwebapp/js/select2_handler.js, line 49 - Missing trailing comma. (comma-dangle)
* corehq/apps/hqwebapp/static/hqwebapp/js/select2_handler.js, line 74 - Missing trailing comma. (comma-dangle)
* corehq/apps/smsbillables/static/smsbillables/js/smsbillables.rate_calc.js, line 20 - 'term' is defined but never used (no-unused-vars)
* corehq/apps/smsbillables/static/smsbillables/js/smsbillables.rate_calc.js, line 23 - Missing trailing comma. (comma-dangle)
* corehq/apps/smsbillables/static/smsbillables/js/smsbillables.rate_calc.js, line 24 - Missing semicolon. (semi)
* corehq/apps/smsbillables/static/smsbillables/js/smsbillables.rate_calc.js, line 46 - Missing trailing comma. (comma-dangle)
* corehq/apps/smsbillables/static/smsbillables/js/smsbillables.rate_calc.js, line 61 - Missing trailing comma. (comma-dangle)
* corehq/apps/smsbillables/static/smsbillables/js/smsbillables.rate_calc.js, line 67 - 'form_data' is defined but never used (no-unused-vars)
* corehq/apps/smsbillables/static/smsbillables/js/smsbillables.rate_calc.js, line 88 - Expected indentation of 12 space characters but found 16. (indent)
* corehq/apps/smsbillables/static/smsbillables/js/smsbillables.rate_calc.js, line 89 - Expected indentation of 12 space characters but found 16. (indent)
* corehq/apps/smsbillables/static/smsbillables/js/smsbillables.rate_calc.js, line 96 - Missing trailing comma. (comma-dangle)

File: corehq/apps/hqwebapp/static/hqwebapp/js/select2_handler.js [hunk 1]
```diff
@@ -1,90 +1,95 @@
...
+        self.getExtraData = function () {
+            return {};
+        };
+    
+        self.processResults = function (response) {
+            // override this if you want to do something special with the response.
+            return response;
+        };
+    
+        self.createNewChoice = function (term, selectedData) {
```
Comment (dimagimon) [inline corehq/apps/hqwebapp/static/hqwebapp/js/select2_handler.js]: 'selectedData' is defined but never used (no-unused-vars)

File: corehq/apps/hqwebapp/static/hqwebapp/js/select2_handler.js [hunk 2]
```diff
@@ -1,90 +1,95 @@
...
+    
+        self.formatResult = function (result) {
+            return result.text;
+        };
+    
+        self.formatSelection = function (result) {
+            return result.text;
+        };
+    
+        self.getInitialData = function (element) {
```
Comment (dimagimon) [inline corehq/apps/hqwebapp/static/hqwebapp/js/select2_handler.js]: 'element' is defined but never used (no-unused-vars)

File: corehq/apps/smsbillables/static/smsbillables/js/smsbillables.rate_calc.js [hunk 3]
```diff
@@ -1,89 +1,90 @@
...
+        self.noError = ko.computed(function () {
+            return ! self.hasError();
+        });
+        self.calculatingRate = ko.observable(false);
+        self.showRateInfo = ko.computed(function () {
+            return self.rate() && ! self.calculatingRate();
+        });
+    
+        self.init = function () {
+            self.select2CountryCode.getExtraData = function (term) {
```
Comment (dimagimon) [inline corehq/apps/smsbillables/static/smsbillables/js/smsbillables.rate_calc.js]: 'term' is defined but never used (no-unused-vars)

File: corehq/apps/smsbillables/static/smsbillables/js/smsbillables.rate_calc.js [hunk 4]
```diff
@@ -1,89 +1,90 @@
...
-    self.showTable = ko.computed(function (){
-        return !self.hasError() && !self.calculatingRate();
-    });
-    self.showRateInfo = ko.computed(function () {
-        return self.rateErrorText() && ! self.calculatingRate();
-    });
-
-    var updateRate = function () {
+    
+    var PublicSMSRateCalculator = function (form_data) {
```
Comment (dimagimon) [inline corehq/apps/smsbillables/static/smsbillables/js/smsbillables.rate_calc.js]: 'form_data' is defined but never used (no-unused-vars)

**Comment by orangejenny** [comment]
ignoring the remaining dimagimon comments

File: corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js [hunk 5]
```diff
@@ -1,437 +1,359 @@
...
-        }
-        return [];
-    });
-    self.hasNoPrivileges = ko.computed(function () {
-        return _.isEmpty(self.selectedPrivileges())
-    });
-};
-
-
-var BaseSelect2Handler = function (options, currentSelectionObj) {
```
Comment (dannyroberts) [inline corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js]: Looks basically like this was copied below with some differences. I don't see `self.currentSelection` in the other one, so am I reading it right that there were no actual usages of that, and in general that the other differences were small enough (or there weren't any) that you can safely replace this implementation with the one below with no real changes?  Kinda wild we had this code around—I guess that's the kind of thing that happens when you don't have a standard way of referencing code between files!

**Review by dannyroberts — APPROVED** [review]
Looks great!

File: corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js (hunk 5 above)
Comment (orangejenny) [inline corehq/apps/accounting/static/accounting/js/accounting.software_plan_version_handler.js]: Yeah, that was my reading - that it was copy-pasted with some tweaks at some point, but the code eventually changed so those tweaks aren't necessary any more.

**Pull request merged**